docs/design-proposals/2022-04 TraceQL Concepts.md

```diff
@@ -0,0 +1,203 @@
+
+### Intrinsic fields
+
+Every span has certain intrinsic fields that can be referenced when selecting spans
+
+| field name | description |
+| ---------- | ----------- |
+| `start`    | start time of the span |
+| `end`      | end time of the span |
+| `duration` | end - start time of the span |
+| `name`     | operation or span name |
+| `status`   | status values are "error", "ok", or "unset" |
+| `parent`   | the parent of this span | 
+
+**Examples**
+Find traces that contain spans whose duration is greater than 2 seconds:  
+`{ duration > 2s }`
+
+Find traces that contain a span named "HTTP POST":  
+`{ name = "HTTP POST" }`
+
+### Attribute fields
+
+We can refer to dynamic attributes (also known as tags) on the span or the span's resource.
+
+**Examples**
+
+Find traces with the GET http method:  
+`{ http.method = "GET" }`
+
+Find traces that passed through the prod namespace:  
+`{ namespace = "prod" }`
+
+Find traces that cross service boundaries:  
+`{ parent.service.name != service.name }`
+
+### Field expressions
+
+Fields can also be combined in various expected ways.
+
+**Examples**
+
+Find traces where the difference of the duration of a set of parent/child spans exceeds 500ms:  
+`{ parent.duration - duration > 500ms }`
+
+Find traces with "success" http status codes:  
+`{ http.status >= 200 && http.status < 300 }`
+
+Note that the second expression requires both conditions to be true on the same span. The entire expression inside of `{}` must be evaluated as true on a single span for it to be included in the resultset.
+
+## Combining Spansets
+
+Logical operators combine sets of spans. For example, if we wanted to find a trace that went through two specific regions:  
+`{ region = "eu-west-0" } && { region = "eu-west-1" } `
+
+Note the difference between the above and the following:  
+`{ region = "eu-west-0"  && region = "eu-west-1" } `
+
+The second expression will return no traces because it's impossible for both conditions to be simultaneously true on the same span.
+
+### Structural operators
+
+Structural operators evaluate traces based on the span relationships within the span tree. 
+
+`{ } >> { }`  
+  This is the descendant operator. The spans returned from this operator will match the right hand side conditions while also being descendants of spans that match the left hand side conditions.
+
+`{ } > { }`  
+  This is the child operator. The spans returned from this operator will match the right hand side conditions while also being children of spans that match the left hand side conditions.
+
+`{ } ~ { }`  
+  This is the sibling operator. The spans returned from this operator will match the right hand side conditions while also being siblings of spans that match the left hand side conditions.
+
+## Aggregators
+
+All of the above expressions involve asking questions about individual spans. However, sometimes we want to ask questions about a set of spans. For that we have aggregate functions.
```

Comment:
Aggregation functionality is something I am missing from APM world as traces/spans seem to be burst data flow. However, when it comes to usage scenarios, it was not about always about attributes of captured span/method/class/etc. In an HTTP scenario, the aggregation key can be query parameters like '?&operation=getItems' or an attribute inside json post body. For even more complex ones, one might need to skip the root or/and a few spans and extract key aggregation parameter in a couple of levels below. 

The good news is, you generally do not need such aggregation for categorization over all nodes transaction visits, rather it gets generally needed on the very first contact point of an application or system.

I do not think  TraceQL should cover all such aspects at once. Just sharing my previous needs in the last decade regarding aggregation.

Reply:
Thanks for the thoughts. Since TraceQL can operate on custom attributes you could extract query params and add them as attributes to the span. Then you would be able to aggregate on them easily. 

We did our best to cover the use cases we found the most important, but of course we will find users with new ideas and needs as the project grows. After we are all using TraceQL please keep bringing these ideas back to the table with real world examples so we can continue to improve the language.